adapter/memQuota/memQuota_test.go

```diff
@@ -0,0 +1,286 @@
+// limitations under the License.
+
+package memQuota
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"istio.io/mixer/pkg/adapter"
+
+	"istio.io/mixer/adapter/memQuota/config"
+	at "istio.io/mixer/pkg/adapter/testing"
+)
+
+type fakeEnv struct {
+	t *testing.T
+}
+
+func (e fakeEnv) Logger() adapter.Logger {
+	return e
+}
+
+func (e fakeEnv) Infof(format string, args ...interface{}) {
+	e.t.Logf(format, args...)
+}
+
+func (e fakeEnv) Warningf(format string, args ...interface{}) {
+	e.t.Logf(format, args...)
+}
+
+func (e fakeEnv) Errorf(format string, args ...interface{}) error {
+	e.t.Logf(format, args...)
+	return fmt.Errorf(format, args)
+}
+
+func TestAlloc(t *testing.T) {
+	definitions := make(map[string]*adapter.QuotaDefinition)
+	definitions["Q1"] = &adapter.QuotaDefinition{
+		MaxAmount: 10,
+		Kind:      adapter.AllocationKind,
+	}
+
+	definitions["Q2"] = &adapter.QuotaDefinition{
+		MaxAmount: 10,
+		Kind:      adapter.PerSecondKind,
+	}
+
+	definitions["Q3"] = &adapter.QuotaDefinition{
+		MaxAmount: 10,
+		Kind:      adapter.PerMinuteKind,
+	}
+
+	b := newBuilder()
+	c := b.DefaultConfig().(*config.Params)
+	c.MinDeduplicationWindow = 3600
+
+	a, err := b.NewQuota(&fakeEnv{t}, c, definitions)
+	if err != nil {
+		t.Errorf("Unable to create aspect")
+	}
+
+	asp := a.(*memQuota)
+
+	cases := []struct {
+		name            string
+		dedup           string
+		allocAmount     int64
+		allocResult     int64
+		allocBestEffort bool
+		tick            int32
+		releaseAmount   int64
+		releaseResult   int64
+	}{
+		{"Q1", "0", 2, 2, false, 0, 0, 0},
+		{"Q1", "0", 2, 2, false, 0, 0, 0}, // should be a nop due to dedup
+		{"Q1", "1", 6, 6, false, 0, 0, 0},
+		{"Q1", "2", 2, 2, false, 0, 0, 0},
+		{"Q1", "3", 2, 0, false, 0, 0, 0},
+		{"Q1", "4", 1, 0, true, 0, 0, 0},
+		{"Q1", "5", 0, 0, true, 0, 0, 0},
+		{"Q1", "5b", 0, 0, false, 0, 5, 5},
+		{"Q1", "5b", 0, 0, false, 0, 5, 5},
+		{"Q1", "5c", 0, 0, false, 0, 15, 5},
+
+		{"Q3", "6", 10, 10, false, 1, 0, 0},
+		{"Q3", "7", 10, 0, false, 2, 0, 0},
+		{"Q3", "8", 10, 0, false, 60, 0, 0},
+		{"Q3", "9", 10, 10, false, 61, 0, 0},
+		{"Q3", "10", 100, 10, true, 121, 0, 0},
+	}
+
+	labels := make(map[string]interface{})
+	labels["L1"] = "string"
+	labels["L2"] = int64(2)
+	labels["L3"] = float64(3.0)
+	labels["L4"] = true
+	labels["L5"] = time.Now()
+	labels["L6"] = []byte{0, 1}
+
+	for i, c := range cases {
+		qa := adapter.QuotaArgs{
+			Name:            c.name,
+			DeduplicationID: "A" + c.dedup,
+			QuotaAmount:     c.allocAmount,
+			Labels:          labels,
+		}
+
+		asp.getTick = func() int32 {
+			return c.tick
+		}
+
+		var amount int64
+		var err error
+
+		if c.allocBestEffort {
+			amount, err = a.AllocBestEffort(qa)
+		} else {
+			amount, err = a.Alloc(qa)
+		}
+
+		if err != nil {
+			t.Errorf("Expecting success, got %v, case %d", err, i)
+		}
+
+		if amount != c.allocResult {
+			t.Errorf("Expecting %d, got %d, case %d", c.allocResult, amount, i)
+		}
+
+		qa = adapter.QuotaArgs{
```

Comment:
I find having it all in one place is more convenient, avoiding a lot of redundancy.